window.c

```diff
@@ -68,10 +79,34 @@ void vtk_window_mainloop(vtk_window win) {
 	XSelectInput(win->root->dpy, win->w, win->event_mask);
 	XMapWindow(win->root->dpy, win->w);
 
+	int x11_fd = ConnectionNumber(win->root->dpy);
+
+	fd_set fds;
+
+	int nfds = win->updatefd;
+	if (x11_fd > nfds) nfds = x11_fd;
+	++nfds;
+
 	XEvent ev;
 	while (!win->should_close) {
-		XNextEvent(win->root->dpy, &ev);
-		vtk_event_handle(win, ev);
+		FD_ZERO(&fds);
+		FD_SET(x11_fd, &fds);
+		FD_SET(win->updatefd, &fds);
+		select(nfds, &fds, NULL, NULL, NULL);
+
+		while (XPending(win->root->dpy) > 0) {
+			XNextEvent(win->root->dpy, &ev);
+			vtk_event_handle(win, ev);
+		}
+
+		uint64_t counter;
+		if (read(win->updatefd, &counter, sizeof counter) != -1) {
+			if (counter > 0 && win->event.update.h) {
+				vtk_event ve;
+				ve.type = VTK_EV_UPDATE;
+				win->event.update.h(ve, win->event.update.d);
```

Comment:
`ve` -> `(vtk_event){ VTK_EV_UPDATE }`

No reason to do something in three lines that's easier to read when done in one. Also it's consistent with how events are dispatched in `event.c`